### PR TITLE
add_progresscheck_tool

### PR DESCRIPTION
**Pull request by duynguyenkos**

### CODE REVIEW

File: ADJ-Internship/BusinessService/Dtos/ProgressCheckDtos.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+        public void CreateMapping(Profile profile)
+        {
+            profile.CreateMap<ProgressCheck, ProgressCheckDto>().IncludeBase<EntityBase, EntityDtoBase>();
+            profile.CreateMap<ProgressCheckDto, ProgressCheck>().IncludeBase<EntityDtoBase, EntityBase>();
+        }
+    }
+    public class GetItemSearchDto : EntityDtoBase
+    {
+        public IEnumerable<string> Origins { get; set; }
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Dtos/ProgressCheckDtos.cs]: don't need an EntityDtoBase for this
EntityDtoBase is just form DTO that maps with Entity

File: ADJ-Internship/BusinessService/Dtos/ProgressCheckDtos.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        public IEnumerable<string> OriginPorts { get; set; }
+        public IEnumerable<string> Factories { get; set; }
+        public IEnumerable<string> Suppliers { get; set; }
+        public IEnumerable<string> Status { get; set; }
+        public IEnumerable<string> Depts { get; set; }
+        
+    }
+    public class OrderDetailDto_Progress : EntityDtoBase, ICreateMapping
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Dtos/ProgressCheckDtos.cs]: Naming Convention, don't use _

Comment (mrkhangtran) [reply ADJ-Internship/BusinessService/Dtos/ProgressCheckDtos.cs]: try another name, OrderProgressDto?

File: ADJ-Internship/BusinessService/Dtos/ProgressCheckDtos.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+        public string Tariff { get; set; }
+
+        [Required]
+        public int OrderId { get; set; }
+
+        public virtual Order Order { get; set; }
+        public void CreateMapping(Profile profile)
+        {
+            profile.CreateMap<OrderDetailDto_Progress,OrderDetail>().IncludeBase<EntityDtoBase, EntityBase>();
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Dtos/ProgressCheckDtos.cs]: why map Progress with Detail?

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ADJ.BusinessService.Implementations
+{
+    public class ProgressCheckService : ServiceBase, IProgressCheckService
+    {
+        private readonly IDataProvider<Order> _odDataProvider;
+        private readonly IOrderRepository _odRepository;
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: make the variable name clearer, _orderDataProvider
or if it is too long, try _progressCheckDataPrd, _progressCheckRepo, ...

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            _odDataProvider = odDataProvider;
+            _odRepository = odRepository;
+
+            _oddRepository = oddRepository;
+            _oddDataProvider = oddDataProvider;
+
+            _prcRepository = prcRepository;
+            _prcDataProvider = prcDataProvider;
+        }
+        public async Task<int> CheckOrderHaventProgress(int orderId)
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: never write negative method
why you return int 0 and 1, is it better to use boolean?
method name should be CheckOrderHasProgress
this is a very bad way to check if the order has progress, find another solution and comment here your new solution

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+                if (item.OrderId == orderId)
+                {
+                    check += 1;                  
+                }
+            }
+            return check;
+        }
+        public void CreateDefaultModel(int orderId)
+        {
+            ProgressCheck progressCheck = new ProgressCheck
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: what is this method for?

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                OrderId = orderId
+            };
+             _prcRepository.Insert(progressCheck);
+            UnitOfWork.SaveChangesAsync();
+
+        }
+        public async Task<PagedListResult<ProgressCheckDto>> ListProgressCheckDtoAsync()
+        {
+            List<ProgressCheckDto> progressCheckDTOs = new List<ProgressCheckDto>();
+            var lstOrder = await _odDataProvider.ListAsync();
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: do not load all these separately
use include in EF to load related object

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+                progressCheckDTOs.Add(temp);
+            }
+
+            PagedListResult<ProgressCheckDto> lstProChDto = new PagedListResult<ProgressCheckDto>
+            {
+                TotalCount = progressCheckDTOs.Count,
+                PageCount = 2,
+                Items = progressCheckDTOs
+            };
+            await UnitOfWork.SaveChangesAsync();
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: this method is getting data, why you save here?

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            return lstProChDto;
+        }
+        public async void Update(ProgressCheckDto progressCheckDTO)
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: never do async void, research more about this

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        {
+            var lstOrder = await _odDataProvider.ListAsync();
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: why updating a progress check needs to load the whole database?

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+            check.EstQtyToShip = (int)temp;
+            if (check.EstQtyToShip == progressCheckDTO.POQuantity)
+            {
+                check.Complete = true;
+            }
+            else
+            {
+                check.Complete = false;
+            }
+            _prcRepository.Update(check);
```
Comment (mrkhangtran) [inline ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: where is save changes?

File: ADJ-Internship/WebApp/WebApp.csproj [hunk 12]
```diff
@@ -30,6 +30,7 @@
   <ItemGroup>
     <ProjectReference Include="..\BusinessService\BusinessService.csproj" />
     <ProjectReference Include="..\Common\Common.csproj" />
+    <ProjectReference Include="..\DataModel\DataModel.csproj" />
```
Comment (mrkhangtran) [inline ADJ-Internship/WebApp/WebApp.csproj]: why you reference DataModel into Web App?

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs (hunk 5 above)
Comment (duynguyenkos) [reply ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: i think i was choose return bool value for this method
        public async Task<bool> CheckOrderHasProgress(int orderId){
                  
} but when i get value for compare then this value is null.

File: ADJ-Internship/WebApp/appsettings.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "DefaultConnection": "Server=.;Database=AdjInternship;Trusted_Connection=True;MultipleActiveResultSets=true;"
+    "DefaultConnection": "Server=DESKTOP0103;Database=AdjInternship;Trusted_Connection=True;MultipleActiveResultSets=true;"
```
Comment (mrkhangtran) [inline ADJ-Internship/WebApp/appsettings.json]: DESKTOP0103 will not work on other machines

File: ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs (hunk 6 above)
Comment (duynguyenkos) [reply ADJ-Internship/BusinessService/Implementations/ProgressCheckService.cs]: when order hasn't progress check i will create for this one default progress check object
.

File: ADJ-Internship/WebApp/Views/ProgressCheck/Index.cshtml [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+                            </td>
+                        </tr>
+                    }
+                </tbody>
+            </table>
+        </div>
+    }
+    <br />
+}
+<script type="text/javascript" src="~/js/jquery-3.3.1.js" asp-append-version="true"></script>
```
Comment (mrkhangtran) [inline ADJ-Internship/WebApp/Views/ProgressCheck/Index.cshtml]: I think this is already in the _Layout

**Pull request merged**